.github/workflows/release.yml

```
@@ -4,21 +4,24 @@ on:
     tags:
       - "*"
 
-permissions:
-  contents: read
+permissions: {}
 
 jobs:
   build:
     name: Build artifacts
+    permissions:
+      contents: read
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8  # v6.0.1
         with:
-          fetch-depth: 0
-      - uses: denoland/setup-deno@v2
+          fetch-depth: 0  # Needed for hatch-vcs versioning
+          persist-credentials: false
+      - uses: denoland/setup-deno@e95548e56dfa95d4e1a28d6f422fafe75c4c26fb  # v2.0.3
         with:
           deno-version: v2.x
```

Better we pin to something in the release workflow at least.
```suggestion
          deno-version: e7b7129b7a92b7500ded88f8f5baa25a7f59e56e
```